File: 1175-prime-arrangements/1175-prime-arrangements.cpp
class Solution {
public:
    int pi(int n) {
        if (n <= 1) {
            return 0;
        }
        int count = 0;
        for (int num=2; num<(n+1); num++) {
            bool has_break = false;
            for (int i=2; i<num; i++) {
                if ((num % i) == 0) {
                    has_break = true;
                    break;
                }
            }
            if (not(has_break)) {
                count++;
            }
        }
        return count;
    }
public:
    double facto(int n, double mod) {
        double res = 1;
        for (int k=2; k<(n+1); k++) {
            res *= k;
            res = std::fmod(res, mod);
        }
        return res;
    }
public:
    double multiplyModulo(int a, int b, double c) {
        double result = 0;
        a = (int) std::fmod((double) a, c);
        b = (int) std::fmod((double) b, c);
        while(b) {
            if (b & 0x1) {
                result += a;
                result = std::fmod(result, c);
            }
            b >>= 1;
            if (a < c - a) {
                a <<= 1;
            } else {
                a -= (c - a);
            }
        }
        return result;
    }
public:
    int numPrimeArrangements(int n) {
        int pi_n = this->pi(n);
        double mod = (pow((double) 10, (double) 9) + 7);
        double res = this->multiplyModulo(this->facto(pi_n, mod), this->facto(n-pi_n, mod), mod);
        return (int) res;
    }
};

File: 1175-prime-arrangements/1175-prime-arrangements.py
class Solution:
    def pi(self, n: int) -> int:
        if n <= 1:
            return 0
        count = 0
        for num in range(2, n+1):
            for i in range(2, num):
                if num % i == 0:
                    break
            else:
                count += 1
        return count
    
    def facto(self, n: int) -> int:
        res = 1
        for k in range(2, n+1):
            res *= k
        return res
    
    def numPrimeArrangements(self, n: int) -> int:
        pi_n = self.pi(n)
        res = self.facto(pi_n) * self.facto(n-pi_n)
        return res % ((10**9 + 7))
